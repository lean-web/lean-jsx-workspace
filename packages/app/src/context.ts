import type { Request } from "express";

export interface RequestQueryParams {
    withLoadingState?: boolean;
    dynamicProductList?: boolean;
    jsDisabled?: boolean;
}

declare module "lean-jsx-types/lib/context" {
    interface SXLGlobalContext extends RequestQueryParams {}
}

export function parseQueryParams(req: Request): RequestQueryParams {
    return {
        jsDisabled: Boolean(req.query?.jsDisabled),
        withLoadingState: Boolean(req.query?.withLoadingState),
        dynamicProductList: Boolean(req.query?.dynamicProductList),
    };
}
