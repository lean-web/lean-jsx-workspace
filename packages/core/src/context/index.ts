interface ComponentContext {
    getId(): string;

    toSource(): string;

    decorate(vnode: SXL.Element): SXL.Element;
}
export type Context<Ctx extends Record<string, unknown>> = Ctx &
    ComponentContext;

export interface SXLGlobalContext {}
