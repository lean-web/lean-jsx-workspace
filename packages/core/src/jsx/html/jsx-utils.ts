export type TextNode = string;
export interface StaticNode extends Omit<SXL.StaticElement, "type"> {
    type: string;
}

export interface FunctionNode extends Omit<SXL.StaticElement, "type"> {
    type: SXL.NodeFactory;
}
export interface ClassNode extends Omit<SXL.StaticElement, "type"> {
    type: SXL.ClassFactory;
}

export function isClass(
    func: string | SXL.NodeFactory | SXL.ClassFactory
): func is SXL.ClassFactory {
    return (
        typeof func === "function" &&
        /^class\s/.test(Function.prototype.toString.call(func))
    );
}

export function isTextNode(jsx: SXL.Element | string): jsx is TextNode {
    return typeof jsx === "string";
}

export function isFunctionNode(
    jsx: SXL.Element | SXL.Children
): jsx is FunctionNode {
    if (typeof jsx === "string" || Array.isArray(jsx)) {
        return false;
    }
    if (isPromise(jsx)) {
        return false;
    }
    return typeof jsx.type === "function" && !isClass(jsx.type);
}

export function isPromise(
    jsx: SXL.StaticElement | Promise<SXL.StaticElement> | undefined | string
): jsx is Promise<SXL.StaticElement> {
    return typeof jsx !== "string" && !!jsx && "then" in jsx;
}

export function isClassNode(jsx: SXL.Element | SXL.Children): jsx is ClassNode {
    if (typeof jsx === "string" || Array.isArray(jsx)) {
        return false;
    }
    if (isPromise(jsx)) {
        return false;
    }
    return typeof jsx.type === "function" && isClass(jsx.type);
}

export function isArrayOfNodes(
    jsx: SXL.StaticElement | SXL.Children
): jsx is Array<SXL.StaticElement | string> {
    return Array.isArray(jsx);
}

export function isStaticNode(
    jsx: SXL.StaticElement | string
): jsx is StaticNode {
    return (
        !isTextNode(jsx) &&
        !isArrayOfNodes(jsx) &&
        !isFunctionNode(jsx) &&
        !isClassNode(jsx)
    );
}

export function isFragmentNode(
    jsx: SXL.Element | undefined
): jsx is FunctionNode {
    return (
        !!jsx &&
        !isPromise(jsx) &&
        typeof jsx.type !== "string" &&
        /^Fragment$/i.test(jsx.type.name)
    );
}

export function unwrapFragments(
    element: SXL.StaticElement | string
): Array<SXL.StaticElement | string> {
    if (typeof element === "string") {
        return [element];
    }
    if (element.type === "fragment") {
        const children = element.children.flatMap((child) =>
            unwrapFragments(child)
        );
        return children;
    }
    if (isFragmentNode(element)) {
        const newElement = element.type({
            ...element.props,
            children: element.children,
        });
        if (isPromise(newElement)) {
            throw new Error("Fragments should not return Promises");
        }
        return unwrapFragments(newElement);
    }
    return [element];
}
