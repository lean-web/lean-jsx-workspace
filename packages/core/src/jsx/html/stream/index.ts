import { Readable, ReadableOptions } from "stream";
import { ContextFactory } from "../context";
import { RenderOptions } from "./stream-api";

import { JSXStream } from "./jsx-stack";

export function getHTMLStreamFromJSX(
    root: SXL.Element,
    contextFactory: ContextFactory,
    htmlTemplate: [string, string],
    options?: ReadableOptions & RenderOptions
): Readable {
    const { jsDisabled, ...opts } = options ?? {};
    const [start, end] = htmlTemplate;

    return new JSXStream(
        root,
        { username: "" },
        { pre: [start], post: [end], sync: jsDisabled ?? false }
    );
}
