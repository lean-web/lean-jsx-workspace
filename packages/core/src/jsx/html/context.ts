/* eslint-disable @typescript-eslint/no-unsafe-call */
import { SXLGlobalContext } from "@/context";
import { isPromise } from "./jsx-utils";
import { UIDGenerator } from "./uuid";

type ExtendedEventHandlers<Ctx> = {
    [k in keyof GlobalEventHandlers]: GlobalEventHandlers[k] extends
        | ((this: GlobalEventHandlers, ev: infer E) => any)
        | null
        ? (this: ExtendedEventHandlers<Ctx>, ev: E, ctx: Ctx) => any
        : never;
};

type ClientHandler<
    T extends keyof ExtendedEventHandlers<Ctx>,
    Ctx
> = ExtendedEventHandlers<Ctx>[T];

export type ContextType = "eager" | "lazy";

interface ComponentContext<Ctx> {
    getId(): string;

    toSource(): string;

    decorate(vnode: SXL.Element): SXL.Element;
}
export type Context<Ctx extends Record<string, unknown>> = Ctx &
    ComponentContext<Ctx>;

type KeyHandlerTyple<Ctx, T extends keyof ExtendedEventHandlers<Ctx>> = [
    T,
    ClientHandler<keyof ExtendedEventHandlers<Ctx>, Ctx>
];

export interface ContextFactoryOptions {
    prefix?: string;
}

export class ContextFactory {
    opts?: ContextFactoryOptions;
    globalContext?: SXLGlobalContext;

    idGen: (type?: "PROD" | "DEV" | undefined) => string;

    constructor(
        opts?: ContextFactoryOptions,
        globalContext?: SXLGlobalContext
    ) {
        this.idGen = UIDGenerator.new();
        this.opts = opts;
        this.globalContext = globalContext;
    }
    newContext<K extends Record<string, undefined>>() {
        const ctx = new ContextImpl<K>(this.idGen());
        const handler1: ProxyHandler<ContextImpl<K>> = {
            get(target, prop, receiver) {
                // console.log({ prop, target });
                if (prop in target) {
                    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
                    return target[prop];
                }
                if (typeof prop === "string") {
                    return target._data[prop];
                }
            },
            set(target, prop, value) {
                if (typeof prop === "string") {
                    target._data[prop] = value;
                    return true;
                }
                return false;
            },
        };

        return new Proxy<ContextImpl<K>>(
            ctx,
            handler1
        ) as unknown as Context<K>;
    }

    buildPlaceholderId(): string {
        const prefix = this.opts?.prefix;
        if (prefix) {
            return `${prefix}-${this.idGen()}`;
        }
        return `placeholder-${this.idGen()}`;
    }
}

export class ContextImpl<Ctx extends Record<string, unknown>>
    implements ComponentContext<Ctx>
{
    private static idGen = UIDGenerator.new();
    private globalContext?: SXLGlobalContext;
    _id: string;
    _handlers: KeyHandlerTyple<Ctx, keyof ExtendedEventHandlers<Ctx>>[] = [];
    _data: Record<string, unknown> = {};

    constructor(id: string, globalContext?: SXLGlobalContext) {
        this._id = id;
        this.globalContext = globalContext;
    }

    getId() {
        return this._id;
    }

    getGlobalContext() {
        return this.globalContext;
    }

    _decorateUnwrappedActions(vnode: SXL.StaticElement) {
        if (!vnode) {
            return;
        }
        const handlers = Object.entries(vnode.props).filter(
            ([key, v]) => /^on/.test(key) || typeof v === "function"
        );

        handlers.forEach(([key, v]) => {
            vnode.props[key] = "";
            // eslint-disable-next-line @typescript-eslint/no-unsafe-argument, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
            this._handlers.push([
                key as keyof GlobalEventHandlers,
                // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                v?.toString(),
            ]);
        });

        return vnode;
    }

    decorate(vnode: SXL.Element): SXL.Element {
        if (!vnode) {
            return vnode;
        }
        if (isPromise(vnode)) {
            return vnode.then((node) => {
                // if (!vnode) {
                //   return vnode;
                // }
                this._decorateUnwrappedActions(node);
                if (this._handlers.length === 0) {
                    return vnode;
                }
                // const dataset = node.props.dataset ?? {};
                if (node.props.dataset) {
                    node.props.dataset[this._id] = "true";
                }
                // node.props.dataset = dataset;
                // console.log({ asyncCtxBefore: node.ctx });
                node.ctx = this as unknown as Context<
                    Record<string, undefined>
                >;
                // console.log({ asyncCtxAfter: node.ctx });
                return node;
            });
        }
        this._decorateUnwrappedActions(vnode);
        if (this._handlers.length === 0) {
            return vnode;
        }
        const dataset = vnode.props.dataset ?? {};
        dataset[this._id] = "true";
        vnode.props.dataset = dataset;
        vnode.ctx = this as unknown as Context<Record<string, undefined>>;

        return vnode;
    }

    toSource() {
        if (this._handlers.length === 0) {
            return "";
        }
        const handlers = this._handlers
            .map(
                ([ev, h]) =>
                    `document.querySelector('[${
                        this._id
                    }]').addEventListener('${ev.replace(
                        /^on/,
                        ""
                    )}', ${h?.toString()})`
            )
            .join(";\n");
        const fns = Object.entries(this._data)
            // .map((entry) => entry[1])
            .filter(
                (entry): entry is [string, Function] =>
                    typeof entry[1] === "function"
            )
            .map(([key, fn]) => `this.${key} =  ${fn.toString()}`)
            .join("\n");
        const source = `<script>
      (function(){
        ${fns}
        ${handlers}
      }).call(${JSON.stringify(this._data)})
    </script>
  `;
        return source;
    }
}
