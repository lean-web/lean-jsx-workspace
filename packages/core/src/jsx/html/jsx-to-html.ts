/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { isStaticNode } from "./jsx-utils";
import { JSDOM } from "jsdom";

function camelToDashed(str: string): string {
    return str.replace(/([a-zA-Z])(?=[A-Z])/g, "$1-").toLowerCase();
}

function normalizeProKey(key: string) {
    return key.replace("className", "class");
}

function normalizePropValue(value: string | number) {
    return typeof value === "number" ? value : `"${value}"`;
}

function styleDeclarationToString(styles: CSSStyleDeclaration): string {
    let styleString = "";
    const entries = Object.entries(styles);
    for (let i = 0; i < entries.length; i++) {
        const [key, value] = entries[i];
        const dashedProperty = camelToDashed(key);
        styleString += `${dashedProperty}: ${value}; `;
    }
    return styleString.trim();
}

function isCSSDeclaration(
    propKey: string,
    propValue: unknown
): propValue is CSSStyleDeclaration {
    return /style/.test(propKey);
}

function flatten(
    props: SXL.Props | Record<string, unknown>
): [string, string | number][] {
    if (Array.isArray(props)) {
        return props;
    }
    return Object.entries(props).flatMap(([key, value]) => {
        if (!value) {
            return [];
        }
        if (/children/.test(key)) {
            return [];
        }
        if (isCSSDeclaration(key, value)) {
            return [[key, styleDeclarationToString(value)]];
        }
        if (typeof value === "object") {
            return flatten(value as Record<string, unknown>);
        }
        if (typeof value !== "string" && typeof value !== "number") {
            throw new Error(
                `Not implemented: Handing property ${key} with value ${value}`
            );
        }
        return [[key, value]];
    });
}

export type OpenAndCloseTags = [string, string];

function serializeObjectToHTML(obj: SXL.StaticElement): [string, string] {
    const { window } = new JSDOM("");
    const { document } = window;

    const { children, ...rest } = obj.props;

    // Create a new element based on the object's tagName property
    const element = document.createElement(obj.type as any);

    // Set attributes from the object
    if (obj.props) {
        for (const [attr, value] of Object.entries(obj.props)) {
            element.setAttribute(attr, value);
        }
    }

    // Get the outerHTML of the created element
    return element.outerHTML.split("><");
}

export class JSXToHTMLUtils {
    static jsxNodeToHTMLTag(jsx: SXL.StaticElement): OpenAndCloseTags {
        if (!isStaticNode(jsx)) {
            throw new Error(
                "Cannot handle JSX nodes with function or class types." +
                    "Please construct the node before calling this method"
            );
        }

        const args = flatten(jsx.props)
            .filter((el) => {
                if (!el) {
                    throw new Error(
                        `Missing prop, ${JSON.stringify(jsx, null, 2)}`
                    );
                }
                return el;
            })
            .map(([key, value]) => {
                try {
                    return `${normalizeProKey(key)}=${normalizePropValue(
                        value
                    )}`;
                } catch (err) {
                    console.error(err, key, value);
                }
            })
            .join(" ");

        return [`<${jsx.type} ${args}>`, `</${jsx.type}>`];

        // const [open, end] = serializeObjectToHTML(jsx);
        // return [`${open}>`, `<${end}`];
    }
}
