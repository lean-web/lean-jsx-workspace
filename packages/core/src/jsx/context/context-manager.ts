import { isPromise, unwrapFragments } from "../html/jsx-utils";
import { UIDGenerator } from "../html/uuid";
import { TrackablePromise } from "../html/stream/stream-utils/trackable-promise";
import { SXLGlobalContext } from "@sxl/core/src/types/context";

interface SyncJSXWrapper {
    id: ContextID;
    element: SXL.StaticElement;
    placeholder: undefined;
    handlers: Array<HandlerPropAndValue>;
    context: SXL.Context<Record<string, unknown>>;
}

interface AsyncJSXWrapper {
    id: ContextID;
    element: TrackablePromise<SXL.StaticElement, unknown>;
    placeholder: SXL.StaticElement;
    handlers: Array<HandlerPropAndValue>;
    context: SXL.Context<Record<string, unknown>>;
}

export type SXLElementWithContext = SyncJSXWrapper | AsyncJSXWrapper;

export function isAsyncElementWithContext(
    element: SXLElementWithContext
): element is AsyncJSXWrapper {
    return (
        element.element instanceof TrackablePromise ||
        isPromise(element.element)
    );
}

export type ContextID = string;

type HandlerPropAndValue = [string, string];

export class LocalContext implements SXL.Context<Record<string, unknown>> {
    [x: string]: unknown;
}

interface ContextManagerOptions {
    sync: boolean;
}

export class ContextManager<G extends SXLGlobalContext> {
    private options: ContextManagerOptions;
    private uiden = UIDGenerator.new();
    private globalContext: G;

    constructor(globalContext: G, options?: ContextManagerOptions) {
        this.globalContext = globalContext;
        this.options = options ?? { sync: false };
    }

    newContext(): SXL.Context<Record<string, unknown>> {
        return new LocalContext();
    }

    fromStaticElement(element: SXL.StaticElement): SXLElementWithContext {
        const id = this.uiden();
        const localCtx = this.newContext();

        return this.processElement(id, localCtx, element);
    }

    fromFunction(element: SXL.FunctionElement): SXLElementWithContext {
        const id = this.uiden();
        const localCtx = this.newContext();

        const newElement = element.type.bind(localCtx)({
            ...element.props,
            children: element.children,
            globalContext: this.globalContext,
        });

        return this.processElement(id, localCtx, newElement);
    }

    fromClass(element: SXL.ClassElement): SXLElementWithContext {
        const id = this.uiden();

        const localCtx = this.newContext();

        const classNode = new element.type({
            ...element.props,
            children: element.children,
            globalContext: this.globalContext,
        });

        return this.processElement(
            id,
            localCtx,
            classNode.renderLazy(),
            classNode.render()
        );
    }

    private processElement(
        id: ContextID,
        context: SXL.Context<Record<string, unknown>>,
        element: SXL.Element,
        placeholder?: SXL.StaticElement
    ): SXLElementWithContext {
        if (isPromise(element)) {
            return {
                id,
                element: new TrackablePromise(
                    element.then((element) => {
                        // if operating in sync mode,
                        // we don't need to wrap the async element
                        // in a template
                        if (this.options.sync) {
                            return element;
                        }
                        if (!element.props.dataset) {
                            element.props.dataset = {};
                        }
                        return {
                            type: "template",
                            props: {
                                id,
                            },
                            children: unwrapFragments(element),
                        };
                    }),
                    id
                ),
                placeholder: this.processPlaceholder(id, placeholder),
                handlers: [],
                context,
            };
        }

        return {
            id,
            element,
            placeholder: undefined,
            handlers: this.processHandlers(id, element),
            context,
        };
    }

    private processHandlers(
        id: ContextID,
        element: SXL.StaticElement
    ): Array<HandlerPropAndValue> {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/ban-types
        const handlers: Array<[string, unknown]> = Object.entries(
            element.props
        ).filter(([key, v]) => /^on/.test(key) || typeof v === "function");

        const _handlers: Array<HandlerPropAndValue> = [];

        handlers.forEach(([key, v]) => {
            let handlerContent = "";
            if (typeof v === "function") {
                handlerContent = v.toString();
            } else if (typeof v === "string") {
                handlerContent = v;
            }
            element.props[key] = "";
            _handlers.push([key as keyof GlobalEventHandlers, handlerContent]);
        });

        if (_handlers.length) {
            element.props.dataset = element.props.dataset ?? {};
            element.props.dataset["data-action"] = id;
        }

        return _handlers;
    }

    private processPlaceholder(
        id: ContextID,
        placehoder?: SXL.StaticElement
    ): SXL.StaticElement {
        return {
            type: "div",
            props: {
                dataset: {
                    ["data-placeholder"]: id,
                },
            },
            children: placehoder ? unwrapFragments(placehoder) : [],
        };
    }
}
