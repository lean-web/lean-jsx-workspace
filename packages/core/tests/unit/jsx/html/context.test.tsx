import { describe, expect, test } from "@jest/globals";
import { ContextFactory } from "@/jsx/html/context";

describe("context", () => {
    function newButton(fn?: () => void): SXL.StaticElement {
        return {
            type: "button",
            props: {
                onclick:
                    fn ??
                    function () {
                        console.log("1");
                    },
            },
            children: [],
        };
    }

    test("context decorates elements with actions correctly", () => {
        const elements: SXL.StaticElement[] = [newButton(), newButton()];
        const factory = new ContextFactory();

        const decorated = elements.map((el) => {
            const ctx = factory.newContext();
            return ctx.decorate(el);
        });

        const ordered = decorated.map(
            (el) => (el as SXL.StaticElement).props.dataset
        );

        expect(ordered[0] && ordered[0]["element-0"]).toBeTruthy();
        expect(ordered[1] && ordered[1]["element-1"]).toBeTruthy();

        expect(decorated).toMatchSnapshot();
    });
});
