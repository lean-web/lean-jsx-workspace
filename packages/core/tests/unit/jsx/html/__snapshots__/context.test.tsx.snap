// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`context context decorates elements with actions correctly 1`] = `
[
  {
    "children": [],
    "ctx": ContextImpl {
      "_data": {},
      "_handlers": [
        [
          "onclick",
          "function () {
                        console.log("1");
                    }",
        ],
      ],
      "_id": "element-0",
      "globalContext": undefined,
    },
    "props": {
      "dataset": {
        "element-0": "true",
      },
      "onclick": "",
    },
    "type": "button",
  },
  {
    "children": [],
    "ctx": ContextImpl {
      "_data": {},
      "_handlers": [
        [
          "onclick",
          "function () {
                        console.log("1");
                    }",
        ],
      ],
      "_id": "element-1",
      "globalContext": undefined,
    },
    "props": {
      "dataset": {
        "element-1": "true",
      },
      "onclick": "",
    },
    "type": "button",
  },
]
`;
